uart: Add basic RS485 mode support

qianfan-Zhao opened the pull request:
this pr is based on: 
dts: set a 'boolean' type's value to 1, not 'True' #8401
extract_dts_include: support 'default' value feature #8463

@erwango There has a simple patch you can used to test this pr. 

this is a simple test patch on nucleo_f070rb platform, RS485's flow control pin is PA1. I had changed uart console to RS485 mode. RS485's flow control pin pull up when sending chars.
```
diff --git a/boards/arm/nucleo_f070rb/nucleo_f070rb.dts b/boards/arm/nucleo_f070rb/nucleo_f070rb.dts
index 72f3e2a..df3317f 100644
--- a/boards/arm/nucleo_f070rb/nucleo_f070rb.dts
+++ b/boards/arm/nucleo_f070rb/nucleo_f070rb.dts
@@ -43,6 +43,8 @@
 	current-speed = <115200>;
 	pinctrl-0 = <&usart2_pins_a>;
 	pinctrl-names = "default";
+	rs485-enabled-at-boot-time;
+	rs485-rts-delay=<10 20>;
 	status = "ok";
 };
 
diff --git a/boards/arm/nucleo_f070rb/pinmux.c b/boards/arm/nucleo_f070rb/pinmux.c
index 96c77e1..7181f84 100644
--- a/boards/arm/nucleo_f070rb/pinmux.c
+++ b/boards/arm/nucleo_f070rb/pinmux.c
@@ -18,6 +18,7 @@ static const struct pin_config pinconf[] = {
 #ifdef CONFIG_UART_STM32_PORT_2
 	{STM32_PIN_PA2, STM32F0_PINMUX_FUNC_PA2_USART2_TX},
 	{STM32_PIN_PA3, STM32F0_PINMUX_FUNC_PA3_USART2_RX},
+	{STM32_PIN_PA1, STM32F0_PINMUX_FUNC_PA1_USART2_RTS},
 #endif	/* CONFIG_UART_STM32_PORT_1 */
 #ifdef CONFIG_I2C_1
 	{STM32_PIN_PB8, STM32F0_PINMUX_FUNC_PB8_I2C1_SCL},
```

Next is waveform captured by logic analyzer: (the first line is RS485's flow control, the second is usart's tx)
![rs4851](https://user-images.githubusercontent.com/16350380/41572498-c1b1c636-73aa-11e8-8b58-aa4f5dfd9c18.png)
![rs4851](https://user-images.githubusercontent.com/16350380/41572605-2ff99bf0-73ab-11e8-86bc-f84432fff430.png)
![rs4851](https://user-images.githubusercontent.com/16350380/41572631-52d51726-73ab-11e8-9d35-552018ac9ae5.png)

Assert time is 10.3us
Deassert time is: 17.1us

qianfan-Zhao committed:
uart: add RS485 mode configure supports

add necessary Kconfig and yaml, also expanded uart_device_config
to store some RS485 configuration variables.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao committed:
stm32f0: add RS485 fixup for stm32f0 series

Add fixup informations for RS485 nodes on STM32F0 series.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao committed:
pinmux: stm32f0: add USART1 and USART2 RTS pinmux

the default pinmux for USART1/2 RTS is: PA1, PA12

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao committed:
driver: stm32-uart: Add RS485 mode support

Some soc supports RS485 hardware flow control using DE pin,
if enabled this feature DE signal is output on the RTS pin.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

zhaozhongchen commented:
@erwango @ydamigos please have a look

ulfalizer commented on arch/arm/soc/st_stm32/stm32f0/dts.fixup:
```diff
@@ -85,4 +85,46 @@
 #define CONFIG_PWM_STM32_17_PRESCALER           ST_STM32_PWM_40014800_PWM_ST_PRESCALER
 
+#ifdef ST_STM32_USART_40013800_RS485_ENABLED_AT_BOOT_TIME
+	#define CONFIG_UART_STM32_USART_1_RS485_ENABLED			1
```
Might want to use some naming other than `CONFIG_...`, to make it clear that these aren't coming from Kconfig.

That might help with finding references to undefined Kconfig symbols later too.

zhaozhongchen replied:
Seems most items in dts.fixup for stm32 series has a CONFIG_,.,, prefix, thats was aren't coming from Kconfig both. I prefer open another PR to remove the CONFIG_ prefix, not is this PR.

What's yours opinion?

ulfalizer replied:
Fine by me, if you'd need to change a bunch of unrelated stuff.

zhaozhongchen replied:
can you write an issue?

ulfalizer replied:
Hmm... not sure how I'd phrase it. Might be other places with non-Kconfig `CONFIG_*` macros as well.

If we add some script for detecting references to undefined Kconfig symbols in source files later, all of them would probably have to be fixed at that point at least.

ydamigos commented on arch/arm/soc/st_stm32/stm32f0/dts.fixup:
```diff
@@ -85,4 +85,46 @@
+#ifdef ST_STM32_USART_40013800_RS485_ENABLED_AT_BOOT_TIME
+	#define CONFIG_UART_STM32_USART_1_RS485_ENABLED			1
+#if !defined(ST_STM32_USART_40013800_RS485_RTS_ACTIVE_LOW)
+	#define CONFIG_UART_STM32_USART_1_RS485_RTS_ACTIVE_LOW		0
```
The fixup file isn't the best place to handle optional properties from DT.

ydamigos reviewed (changes requested):
The driver should handle the optional properties from DT. Keep in mind that the dts.fixup files will be removed in the future.

zhaozhongchen commented:
@ydamigos 

> The fixup file isn't the best place to handle optional properties from DT.

Yes I know but I can't find a better way.

rs485-enabled-at-boot-time is a 'boolean' type, if it is exist in dts, there has sometings like this:
```
#define ST_STM32_USART_40004400_RS485_ENABLED_AT_BOOT_TIME		True
```

But if you haven't set rs485-enabled-at-boot-time, DTS generater doesn't define 'ST_STM32_USART_40004400_RS485_ENABLED_AT_BOOT_TIME'. So I need handler thats by myslef.

As the same rs485-rts-delay is an array if not defined we need keep the default value <0 0>, but the DTS generater doesn't has a default value feature.

So I need a way to handler this, I don't want handle on drivers, so I just let them in dts.fixup.
Is there has another better way to handle this?

qianfan-Zhao committed:
dts: set a 'boolean' type's value to 1, not 'True'

Fixes: #8376

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

qianfan-Zhao committed:
extract_dts_include: support 'default' value feature

Insert a device tree node with default value based on YAML if
that's node doesn't present in dts.

This is a example for uart.yaml:
rs485-enabled-at-boot-time:
  type: boolean
  default: True
  category: optional
  description: Enable RS485 mode default
  generation: define

if rs485-enabled-at-boot-time doesn't present in dts, will generate
a deault value(True) for *all* uart devices.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

zhaozhongchen commented:
@erwango @ydamigos I had updated, please recheck

erwango reviewed (changes requested):
Please provide a sample so we can see how this all interacts and one can actually test.
